[StdLib] Changed interface to mutual term-level recursion

effectfully opened the pull request:
I moved some stuff that handles mutually recursive functions from `plutus-ir` to `language-plutus-core` which allows to greatly simplify tests in `language-plutus-core/test/Evaluation/CkMachine.hs`.

While thinking how to that nicely, I realized that there is a simpler way to handle n-ary tuples. Before, tuple types, tuple constructors and accessors were all closed and it was necessary to perform n-ary applications/instantiations over them. But when we take the fixed point of a family of functions, all the types are already instantiated in the result and we can generate tuple accessors without first abstracting types and then instantiating them. I.e. we can treat a tuple as a pair of

1. types of the elements of the tuple
2. the actual term

The benefit is that both `language-plutus-core` code that handles mutually recursive functions becomes much simpler. E.g. here is how we now define `getEven` in terms of `getEvenAndOdd`:

```haskell
getEven = getEvenAndOdd >>= tupleTermAt () 0
```

The code in `plutus-ir` becomes simpler as well, but mostly because stuff is abstracted into `language-plutus-core`.

The old tuple handling function are no longer needed and I commented them out, however they make perfect sense and we may not want to remove them. I could try to define old functions in terms of the new ones. What do you think?

Don't merge yet, I'll add some docs.

michaelpj commented on language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Function.hs:
```diff
@@ -247,3 +250,44 @@ getBuiltinFixN n = do
+    let chooseTy = mkIterTyFun ann funTys (TyVar ann q)
+
+    -- \v1 ... vn -> choose f1 ... fn
+    vsLam <- do
```
No need for this to be a do block, could just be a nested let now.

michaelpj commented on language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Function.hs:
```diff
@@ -247,3 +250,44 @@ getBuiltinFixN n = do
         LamAbs () h hTy $
         mkIterApp () (Var () h) branches
+
+data Function funName ann = Function
```
Maybe put this and the two below in `MkPlc`? Seems useful in a similar way to `VarDecl` and co.

michaelpj replied:
Or maybe we need more stuff in a meta-PLC lib between the main one and the stdlib.

michaelpj commented on language-plutus-core/stdlib/Language/PlutusCore/StdLib/Meta/Data/Tuple.hs:
```diff
@@ -3,104 +3,163 @@
+--         -- \arg_1 .. arg_n . arg_i
+--         mkIterLamAbs () args selectedArg
+
+data Tuple ann = Tuple
```
Well, this is the old discussion about how much should be meta and how much should be in PLC. The previous approach of this module was to have the minimum meta - only the arity, which would have been quite annoying to do in PLC. This puts much more into meta. Not sure which is better tbh.

effectfully replied:
> The previous approach of this module was to have the minimum meta - only the arity

Library-wise, yes. But when we actually need to use that library, we have to keep the necessary meta-information in order to instantiate type abstractions.

So we have that `getBuiltinMutualFixOf` function. It returns the fixed point of a family of mutually recursive functions as a Scott-encoded tuple. But if we kept only the term, it would mean that we can't actually do much with the tuple. For example, you can't really take the ith element of a tuple without knowing the types of its elements. Of course, you can write the generic accessor function like the one we had before, but this is just asking the user to keep track of types and provide them later. Those are details that we can remember and free the user from them: `getBuiltinMutualFixOf` returns an encapsulated tuple, elements of which can be accessed without the need to explicitly pass types around.

Previously it was

```
    user
  /      \
tuple  element types
  \      /
   result
```

now it's

```
    user
     |
   tuple
     |
   result
```

But this is "theoretically". Practically, there is not much difference. The user still has to supply names to `bindTuple` (tuples are anonymous and it doesn't make sense to hardcode names for elements) where he previously need to supply typed names. I anyway think this is a more fine-grained solution, because previously we were forcing the user to supply types twice: when constructing a tuple and when accessing it, while now we require to supply types only once. And the new solution allows to cut some corners and generate less type abstractions/instantiations and is also much more convenient when you write Plutus Core by hand.

michaelpj replied:
It would be good to have functions around for creating the tuple types, constructors etc. themselves, since that can be handy if you actually want to pass that as an argument or something.

michaelpj commented on language-plutus-core/stdlib/Language/PlutusCore/StdLib/Meta/Data/Tuple.hs:
```diff
@@ -3,104 +3,163 @@
+
+    pure
+        ( selectedTy
+        , Apply ann (TyInst ann term selectedTy) $ mkIterLamAbs ann args selectedArg
```
This is going to duplicate the entire tuple every time you access an element, no?

effectfully replied:
Yes, just like an application of `getBuiltinTupleAccessor` did previously. This is not a problem as long as `term` is a variable. I.e. this is no different from the previous solution, but needs to be documented properly.

michaelpj commented on language-plutus-core/test/Evaluation/CkMachine.hs:
```diff
@@ -7,172 +7,91 @@ import           Language.PlutusCore
-        li1 <- listIntToListInt
-        li2 <- listIntToListInt
-        pure $ TyFun () li1 (TyFun () li2 (TyVar () q))
+getEven = getEvenAndOdd >>= tupleTermAt () 0
```
This *is* nice, but I think it's better to have abstraction in the generated code too. It does make accessing it more awkward, I admit. But I think we could still do something like this, but just generate the appropriately instantiated tuple accessor.

effectfully replied:
> This is nice, but I think it's better to have abstraction in the generated code too.

For sure. It was just a stupid bug that caused tuple duplication.

> But I think we could still do something like this, but just generate the appropriately instantiated tuple accessor.

An application of an accessor still duplicates a tuple. What matters duplication-wise is whether a tuple is a variable or really an inlined tuple. Whether you define an accessor and then apply it to a tuple or just apply the tuple directly to a selector is irrelevant, because you're going to duplicate the tuple in any case (which is fine if the tuple is a variable).

michaelpj replied:
Right, fair.

effectfully commented on plutus-ir/test/recursion/even3.plc.golden:
```diff
@@ -1,79 +1,609 @@
 (program 1.0.0
```
I thought of the problem of not duplicating tuples, solved it and then stupidly forgot to implement the solution. Thanks for catching this. Updated the file, it's now shorter than it was originally, because we do not generate unnecessary instantiations.

effectfully commented on language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Function.hs:
```diff
@@ -247,3 +250,44 @@ getBuiltinFixN n = do
+    let funTys = map functionType funs
+
+    q <- liftQuote $ freshTyName ann "Q"
+    -- TODO: It was 'safeFreshName' previously. Should we perhaps have @freshName = safeFreshName@?
```
> uncontrolled input.

I have no trust in my fingers. I think it's better to be extra paranoid (always a good thing) and provide a uniform API (always a good thing as well). But that's for another PR anyway.

michaelpj replied:
:+1:

michaelpj commented on language-plutus-core/stdlib/Language/PlutusCore/StdLib/Data/Function.hs:
```diff
@@ -247,3 +250,42 @@ getBuiltinFixN n = do
+functionVarDecl :: Function (Name ann) ann -> VarDecl TyName Name ann
+functionVarDecl fun@(Function ann _ _ (Def name _)) = VarDecl ann name $ functionType fun
+
+getBuiltinMutualFixOf :: ann -> [Function (Name ann) ann] -> Quote (Tuple ann)
```
This is now also quite "meta". I only mention this because we do distinguish them in the module structure, but I'm not sure how useful that is.

michaelpj replied:
Sure, just good to bear in mind.

effectfully commented:
Yeah, thanks, I'm addressing comments.

effectfully commented:
@michaelpj, please take another look. I renamed `Function` to `FunctionDef` and moved it to`MkPlc.hs`, added comments and did some other minor administrative work. I kept your stuff from `Tuple.hs`, but renamed there `Tuple` to `ProdN`, because those are proper n-ary products, while the new stuff is all about term-level tuples ("tuple" has more "programming" connotations, while "n-ary product" has more theoretic annotations).

effectfully commented:
Takes more than hour to run tests?

If going to merge, please squash. I tried to do this locally, but only did mess. I don't know why I'm unable to squash when there are merge commits.

michaelpj commented:
It's passed, it seems to be not reporting the status again. 

For squashing, I usually find you need to rebase on master so the merge commits drop out, and then squash things. Might be an easier way.
